src/utils/typedData.ts

```diff
@@ -107,130 +174,218 @@ function getMerkleTreeType(types: TypedData['types'], ctx: Context) {
-    });
-    const { root } = new MerkleTree(structHashes as string[]);
-    return ['felt', root];
-  }
+  if (type.endsWith('*')) {
+    const isStructArray = Object.keys(types)
+      .map((x) => `${x}*`)
+      .includes(type);
 
-  if (type === 'felt*') {
-    return ['felt*', computeHashOnElements(data as string[])];
-  }
+    const mappingMethod: (entry: Record<string, unknown>) => string = isStructArray
+      ? (entry) => getStructHash(types, type.slice(0, -1), entry, revision)
+      : (entry) => encodeValue(types, type.slice(0, -1), entry, undefined, revision)[1];
 
-  if (type === 'selector') {
-    return ['felt', prepareSelector(data as string)];
+    const hashes: string[] = (data as Array<TypedData['message']>).map(mappingMethod);
+    return [type, revisionConfiguration[revision].hashMethod(hashes)];
   }
 
-  return [type, getHex(data as string)];
-};
+  switch (type) {
+    case 'enum': {
+      if (revision === Revision.Active) {
+        const [variantKey, variantData] = Object.entries(data as TypedData['message'])[0];
+
+        const parentType = types[ctx.parent as string][0] as StarkNetEnumType;
+        const enumType = types[parentType.contains];
+        const variantType = enumType.find((t) => t.name === variantKey) as StarkNetType;
+        const variantIndex = enumType.indexOf(variantType);
+
+        const encodedSubtypes = variantType.type
+          .slice(1, -1)
+          .split(',')
+          .map((subtype, index) => {
+            if (!subtype) return subtype;
+            const subtypeData = (variantData as unknown[])[index];
+            return encodeValue(types, subtype.trim(), subtypeData, undefined, revision)[1];
+          });
+        return [
+          type,
+          revisionConfiguration[revision].hashMethod([variantIndex, ...encodedSubtypes]),
+        ];
+      } // else fall through to default
+      return [type, getHex(data as string)];
+    }
+    case 'merkletree': {
+      const merkleTreeType = getMerkleTreeType(types, ctx);
+      const structHashes: string[] = (data as Array<TypedData['message']>).map((struct) => {
+        return encodeValue(types, merkleTreeType, struct, undefined, revision)[1];
+      });
+      const { root } = new MerkleTree(structHashes as string[]);
+      return ['felt', root];
+    }
+    case 'selector': {
+      return ['felt', prepareSelector(data as string)];
+    }
+    case 'string': {
+      if (revision === Revision.Active) {
+        const byteArray = byteArrayFromString(data as string);
+        const elements = [
+          byteArray.data.length,
+          ...byteArray.data,
+          byteArray.pending_word,
+          byteArray.pending_word_len,
+        ];
+        return [type, revisionConfiguration[revision].hashMethod(elements)];
+      } // else fall through to default
+      return [type, getHex(data as string)];
+    }
+    case 'felt':
+    case 'bool':
```

Comment:
could have some check for booleans to avoid passing things like "3" here

Reply:
Will be added for v6 along with boundary checks for the other base types.